### PR TITLE
Add Inject instance for tuples

### PR DESCRIPTION
**Pull request by bosu**
`(a, b)` is mapped to `{ _1 = a, _2 = b }`

### CODE REVIEW

**Comment by Gabriel439** [comment]
What I'd suggest doing here is to first define an `R2` type, like this:

```haskell
data R2 a b = R2 { _1 :: a, _2 :: b }
    deriving (Generic, Inject, Interpret)
```

... and then defining the `Inject` instance for the tuple by taking advantage of the `Contravariant` instance for `InputType`:

```haskell
instance (Inject a, Inject b) => Inject (a, b) where
    injectWith = fmap (contraMap adapt) injectWith
      where
        adapt (_1, _2) = R2 {..})
```

The advantage of this approach is that it's easy to scale out to larger tuple sizes

The other advantage is that you can also export the `R{N}` family of constructors so that users can also use those directly instead of tuples if they want the correspondence to the Dhall type to be more clear

**Comment by Gabriel439** [comment]
Also, while you're doing this, you can also define a matching `Interpret` instance for tuples, too

**Comment by bosu** [comment]
> What I'd suggest doing here is to first define an R2 type, like this:

You are right, this looks much better. I'll update the pull request.

> Also, while you're doing this, you can also define a matching Interpret instance for tuples, too

Will do.

**Comment by bosu** [comment]
Rebased and updated the pull request.

File: src/Dhall.hs [hunk 1]
```diff
@@ -103,6 +105,12 @@ instance Show InvalidType where
 
 instance Exception InvalidType
 
+-- | Dhall uses this type to represent (a, b) tuple.
+data R2 a b = R2 { _1 :: a, _2 :: b } deriving (Show, Generic)
+
+instance (Inject a, Inject b) => Inject (R2 a b)
```
Comment (Gabriel439) [inline src/Dhall.hs]: If you add the `DeriveAnyClass` extension, then you can simplify this further as just:

```haskell
data R2 a b = R2 { _1 :: a, _2 :: b }
    deriving (Generic, Inject, Interpret, Show)
```

**Comment by bosu** [comment]
Update pull request uses `DeriveAnyClass`

**Comment by Gabriel439** [comment]
I'll go ahead and merge this and then fix the build failure on `master` separately since this pull request is not the cause of the problem

**Pull request merged**